### PR TITLE
Do not balance partial searchable snapshots by disk

### PR DESCRIPTION
**Pull request by idegtiarenko**
All partial searchable snapshot indices are not restored fully but instead rely on caching recently read segments.
Such indices are marked with the following settings:
https://github.com/elastic/elasticsearch/blob/fe0b1748d2e7021d1ca1349d96aaea7232acbfcd/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java#L142-L147

As a result they are reported with 0 size in the ClusterInfo, see
https://github.com/elastic/elasticsearch/blob/000ffaae05bdd7352542168e278f427bc871a12d/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1334-L1337

Since we do not clear `org.elasticsearch.cluster.metadata.IndexMetadata#shardSizeInBytesForecast` balancer was falling back to its value and was balancing frozen indices by it. 

This commit assigns 0 disk usage during balancing to avoid unnecessary frozen indices sahrds movements

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -371,6 +371,9 @@ private static long getTotalDiskUsageInBytes(ClusterInfo clusterInfo, Metadata m
 
         // Visible for testing
         static long getIndexDiskUsageInBytes(ClusterInfo clusterInfo, IndexMetadata indexMetadata) {
+            if (indexMetadata.ignoreDiskWatermarks()) {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:374]: Can we add a comment that mentions that we ignore disk watermarks for partially mounted/frozen indices?

**Review by fcofdez — APPROVED** [review]
LGTM

**Review by DaveCTurner — APPROVED** [review]
LGTM2

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-4

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java [hunk 2]
```diff
@@ -269,6 +260,26 @@ public void testBalanceByActualAndForecastDiskUsage() {
...
+            stateWithStartedIndices(
+                anIndex("frozen-index-1", partialSearchableSnapshotSettings).shardSizeInBytesForecast(ByteSizeValue.ofGb(1).getBytes()),
+                anIndex("frozen-index-2", partialSearchableSnapshotSettings).shardSizeInBytesForecast(ByteSizeValue.ofGb(1).getBytes()),
+                anIndex("frozen-index-3", partialSearchableSnapshotSettings).shardSizeInBytesForecast(ByteSizeValue.ofGb(1).getBytes()),
+                anIndex("frozen-index-4", partialSearchableSnapshotSettings).shardSizeInBytesForecast(ByteSizeValue.ofGb(10).getBytes())
+            ),
+            allocationService
+        );
+
+        assertThat(getShardsPerNode(clusterState).values(), everyItem(hasSize(2)));
```
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:280]: BTW, what are these two items? A comment may be helpful. My guess is that one is the with the zero searchable snapshots?

**Review by kingherc — APPROVED** [review]
LGTM3

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java (hunk 2 above)
Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java:280]: All indices are partial searchable snapshots here. We do not care where each index is allocated as long as they are balanced by count across the nodes

**Pull request merged**